File: LeetCodeProblemsGo/LeetCode955DeleteColumnsToMakeSortedII/LeetCode955DeleteColumnsToMakeSortedII.go
package LeetCode955DeleteColumnsToMakeSortedII

func minDeletionSize(strs []string) int {
	height := len(strs)
	width := len(strs[0])
	columnsToDelete := 0
	rowsToCheck := make(map[int]any)
	for i := 0; i < height-1; i++ {
		rowsToCheck[i] = nil
	}
	colDeleted := false
	newRowsToCheck := make(map[int]any)
	for col := 0; col < width; col++ {
		colDeleted = false
		newRowsToCheck = make(map[int]any)
		for key, _ := range rowsToCheck {
			newRowsToCheck[key] = nil
		}
		for row := 0; row < height-1; row++ {

			if _, exists := rowsToCheck[row]; !exists {
				continue
			}
			if strs[row][col] > strs[row+1][col] {
				columnsToDelete++
				colDeleted = true
				break
			}
			if strs[row][col] < strs[row+1][col] {
				delete(newRowsToCheck, row)
			}
		}
		if !colDeleted {
			rowsToCheck = make(map[int]any)
			for key, _ := range newRowsToCheck {
				rowsToCheck[key] = nil
			}
		}
	}
	return columnsToDelete
}
func Calculate() {
	strs := []string{"vdy", "vei", "zvc", "zld"}
	result := minDeletionSize(strs)
	println(result)
}

File: LeetCodeProblemsCMake/Problems/LeetCode955DeleteColumnsToMakeSortedII.cpp
//
// Created by saeed on 12/21/25.
//

#include "LeetCode955DeleteColumnsToMakeSortedII.h"
using namespace std;

int LeetCode955DeleteColumnsToMakeSortedII::minDeletionSize(vector<string> &strs) {
    int height=static_cast<int>(strs.size());
    int width=static_cast<int>(strs[0].size());
    int colsToDelete=0;
    vector<bool> rowsToCheck(height-1,true);
    vector<bool> newRowsToCheck(height-1,true);

    bool colDeleted=false;

    for (int col = 0; col < width; ++col) {
        colDeleted = false;
        newRowsToCheck = rowsToCheck;

        for (int row = 0; row < height - 1; ++row) {
            if (!rowsToCheck[row])
                continue;

            if (strs[row][col] > strs[row + 1][col]) {
                ++colsToDelete;
                colDeleted = true;
                break;
            }

            if (strs[row][col] < strs[row + 1][col]) {
                newRowsToCheck[row] = false;
            }
        }

        if (!colDeleted) {
            rowsToCheck = newRowsToCheck;
        }
    }
    return colsToDelete;
}
